let's make a database/sql driver

alicebob opened the pull request:
Shouldn't be too hard to get a (readonly) database/sql driver working which supports basic selects. Right?

alicebob committed:
driver skeleton

alicebob committed:
comments

alicebob committed:
basic tablescan

vtolstov commented:
very cool!

sbinet commented on driver/driver.go:
```diff
@@ -0,0 +1,178 @@
+type Statement struct {
+	dbh *db.Database
+	SQL string
+}
```
```suggestion
}

var (
    _ driver.Stmt = (*Statement)(nil)
)
```

just to make sure?

alicebob replied:
Added, thanks!

alicebob committed:
'*' is a special token

alicebob committed:
parse the query and get the tablename from the SELECT

If it's a SELECT, otherwise we'll complain.

alicebob committed:
Update driver/driver.go

Co-Authored-By: Sebastien Binet <binet@cern.ch>

alicebob committed:
use the current Go version

alicebob committed:
add db.Columns() method

alicebob committed:
use the main sqlittle to access the tables

It already has all the details about columns and rows, which can get
weird sometimes.

alicebob committed:
don't pretend Exec() works

alicebob committed:
support column names in SELECT

It doesn't yet error on invalid column names.

greenpau commented:
This is great! @alicebob , how far are you from the finish line on this one?

alicebob commented:
On Wed, Apr 29, 2020 at 06:32:35PM -0700, Paul Greenberg wrote:
> This is great! @alicebob , how far are you from the finish line on this one?

I remember two problems:
- cancelling a query after fetching some, but not all, rows
- getting "unknown column" errors already in the sql.Query() call, not on the
  first Scan().

Both don't fit easily in the way sqlittle works. But they aren't impossible
to do either.

alicebob committed:
introduce `SelectDone()`, which can stop iterating

alicebob committed:
make rows.Close() stop the iterating, no go routine leak

alicebob committed:
don't drop the error for invalid column names

"invalid column" errors will only be returned via rows.Err() for now.

alicebob committed:
implement QueryContext instead

alicebob committed:
test for QueryRow

Nothing we need to do to support that function, but just a test to be
sure it really works.

alicebob committed:
no more autoreadme :(

It doesn't support Go modules. Any alternative?

alicebob committed:
add driver to the main README

alicebob committed:
Merge branch 'master' into driver

alicebob committed:
comment typo

alicebob committed:
be more explicit about the implemented interfaces

alicebob commented:
This is Good Enough as a first version. Next up would be better support for data types (I guess mimic what the main sqlite driver does), and then maybe add some simple SQL features.
